packages/auth/src/authClients/firebase.ts

```diff
@@ -14,118 +18,130 @@ export type oAuthProvider =
   | 'microsoft.com'
   | 'apple.com'
 
-export type Prompt = 'none' | 'consent' | 'select_account'
-
-// valid parameters as of 2021-06-12 at https://firebase.google.com/docs/reference/js/firebase.auth.GoogleAuthProvider#setcustomparameters
-export type CustomParameters = {
-  hd?: string
-  include_granted_scopes?: boolean
-  login_hint?: string
-  prompt?: Prompt
-}
+export type emailLinkProvider = 'emailLink'
 
 export type Options = {
-  providerId?: oAuthProvider
+  providerId?: oAuthProvider | emailLinkProvider
   email?: string
+  emailLink?: string
   password?: string
   scopes?: string[] // scopes available at https://developers.google.com/identity/protocols/oauth2/scopes
-  customParameters?: CustomParameters
+  customParameters?: CustomParameters // parameters available at https://firebase.google.com/docs/reference/js/firebase.auth.GoogleAuthProvider#setcustomparameters
 }
 
 const hasPasswordCreds = (options: Options): boolean => {
   return options.email !== undefined && options.password !== undefined
 }
 
-export const firebase = (client: Firebase): AuthClient => {
-  // Use a function to allow us to extend for non-oauth providers in the future
-  const getProvider = (providerId: oAuthProvider) => {
-    return new client.auth.OAuthProvider(providerId)
+const applyProviderOptions = (
+  provider: OAuthProvider,
+  options: Options
+): OAuthProvider => {
+  if (options.customParameters) {
+    provider.setCustomParameters(options.customParameters)
   }
-  const applyProviderOptions = (provider: any, options: Options) => {
-    if (options.customParameters) {
-      provider.setCustomParameters(options.customParameters)
-    }
-    if (options.scopes) {
-      options.scopes.forEach((scope) => provider.addScope(scope))
-    }
-    return provider
+  if (options.scopes) {
+    options.scopes.forEach((scope) => provider.addScope(scope))
   }
-  // Firebase auth functions return a goog.Promise which as of 2021-05-12 does
-  // not appear to work with try {await} catch blocks as exceptions are not caught.
-  // This client returns a new standard Promise so that the exceptions can be
-  // caught and no changes are required in common auth code located in AuthProvider.tsx
-  const repackagePromise = (
-    fireBasePromise: Promise<any | string>
-  ): Promise<any | string> => {
-    return new Promise((resolve, reject) => {
-      fireBasePromise
-        .then((result) => resolve(result))
-        .catch((error) => reject(error))
-    })
+  return provider
+}
+
+type FirebaseClient = {
+  firebaseAuth: FirebaseAuth
+  firebaseApp?: FirebaseApp
+}
+
+export const firebase = ({
+  firebaseAuth,
+  firebaseApp,
+}: FirebaseClient): AuthClient => {
+  const auth = firebaseAuth.getAuth(firebaseApp)
```

Comment:
Hey @doesnotexist - could you help me understand something? 

Why do we need to do this, and whats the difference between the object returned from getAuth, and the imported `import * as firebaseAuth from '@firebase/auth'`?

Reply:
This is done so that firebase is not a dependency for normal redwood projects that are not using the firebase auth provider.  Instead, users who are using firebase auth, are required to pass the module to this factory function so it can call the associated functions.  One export is the getAuth() method which returns the auth object (or creates and initializes an auth object) for the firebase app or the default firebase app if no app is given.   In VS Code, it might help to look at the autocomplete suggests after typing firebaseAuth. 

The reason this looks so much different than before has to do with the firebase v8 to v9 transition, to support tree-shaking.  Previously there was one monolithic firebase import and to get the auth instance you called a method of the firebase instance:
firebaseApp.auth() 
is the equivalent to getAuth() which is now a module export in the @firebase/auth module.

One thing I think is confusing about what I've written is, though similarly named there is a distinction between the two properties in the FirebaseClient type,  firebaseApp is an instance method created by initializeApp() which is more analogous to getAuth() function but for the top-level firebase module import,  wherease firebaseAuth is the module exports object and not an instance, which is created by calling firebaseAuth.getAuth() using the firebaseAuth module passed in by the user.

Reply:
Also, I should mention that isn't sufficient to just pass the auth instance into this factory function as there are multiple other functions needed from the module that no longer live as methods on an auth instance but are just exports from the module.  so even if we passed an auth instance we'd still need to pass in the modules export module in order to call those other functions in the module exports object that are no longer methods of an auth instance.

Reply:
I tried something else see commit here: https://github.com/redwoodjs/redwood/commit/a47974ea3e8a8a2f1e26630a6d627a4cfbe0f6f8

And we pass it to the auth provider like so:

```js
const firebaseApp = ((config) => {
  const apps = getApps()
  if (!apps.length) {
    initializeApp(config)
  }
  return getApp()
})(firebaseConfig)

const App = () => (
//...
    <RedwoodProvider titleTemplate="%PageTitle | %AppTitle">
      <AuthProvider client={firebaseApp} type="firebase">
```

What do you think?

Reply:
Also I worked out the difference - getAuth gives you an instanceon the auth client. Where as the imports are just utility functions, so maybe there's no need to pass it through from the app level right?

Reply:
I also prefer a47974e and I tried similar with 70c59e but had to revert it back to passing the module exports because when I did create-redwood-app and then 'yarn rw dev' the default app had a compile error.  
Described here: https://github.com/redwoodjs/redwood/pull/3347#issuecomment-923366921

Have you tried create-redwood-app with your changes?  I think you might run into the same issue

Also, early on I was passing an object with only those exported utility functions before, which might be more ideal from tree-shaking but required a new typescript type to describe the subset of functions required from @firebase/auth's modules.export ... But perhaps thats ok, I was a less familiar with typescript back when I was doing that.